### DISCUSSION TITLE
Convergence problems in solving Poisson equation with Poisson-Boltzmann relation

### DISCUSSION THREAD

**Discussion by ahzeeshan** [idaholab/moose#26561] [Q&A General] [answered]
I am trying to solve the Poisson's equation for electrical potential in a system containing positively and negatively charged ions,
$$\nabla ^2 \phi = - \rho/ \epsilon = - e(c^+ - c^-)/\epsilon$$
where $c^+$, $c^-$ are the concentrations given by
$$c^{\pm} = N \exp\left[\frac{ A^{\pm} \pm e\phi }{kT}  \right]$$
where $A^{\pm}, N, kT, e, \epsilon$ are constants.

I used the `Diffusion` and the `MaskedBodyForce` kernels (in the phase field module) for `phi`. The `MaskedBodyForce` is used for the $-\rho/\epsilon$ term. I am getting a convergence problem: `Linear solve did not converge due to DIVERGED_BREAKDOWN iterations`

Since the problem is very non-linear, I used a gradual ramp-up with time for the `MaskedBodyForce` term by defining `function = 't/10'` and running the simulation till `end_time=10.0`. This does not converge beyond the first iteration. The input and output files are in https://github.com/ahzeeshan/moose-issue/tree/main/poisson

I am using similar values for the constants as a paper where they solved the equation in MATLAB.

I also wanted to confirm: the material `m` in the [`MaskedBodyForce`](https://mooseframework.inl.gov/source/kernels/MaskedBodyForce.html) can be a function of the variable `phi`, right?

**Comment by lindsayad** [comment]
Hi, the preconditioning matrix seems to be very poor. You can check this by decreasing the size of the problem and running `-snes_test_jacobian`. I was able to get some convergence by switching to PJFNK. My edits to your input are below. I see that you're using the `DerivativeMaterialInterface` so I wouldn't expect your matrix to be so poor, but I don't really use that feature much myself. Perhaps @dschwen could see what's wrong at a glance
```
diff --git a/modules/phase_field/user-problem.i b/modules/phase_field/user-problem.i
index 16bb511818..318ec1186b 100644
--- a/modules/phase_field/user-problem.i
+++ b/modules/phase_field/user-problem.i
@@ -2,7 +2,7 @@
   [./mesh]
    dim = 1
    type = GeneratedMeshGenerator
-   nx = 100000
+   nx = 10000
    xmax = 100e-9
   [../]
 []
@@ -119,18 +119,19 @@
   type = Transient
   scheme = bdf2
   verbose = True
-  solve_type = 'Newton'
-  l_max_its = 50
-  l_tol = 1e-6
-  nl_max_its = 50
+  solve_type = 'PJFNK'
+  l_max_its = 300
+  l_tol = 1e-2
+  abort_on_solve_fail = true
+  nl_max_its = 30
   nl_rel_tol = 1e-8
   nl_abs_tol = 1e-12
 #  petsc_options = '-snes_monitor -ksp_monitor_true_residual -snes_converged_reason -ksp_converged_reason'
 #  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
 #  petsc_options_value = 'asm      31                  preonly      lu          4'
 #  line_search = 'none'
-  petsc_options_iname = -pc_type
-  petsc_options_value = asm
+  petsc_options_iname = '-pc_type -ksp_gmres_restart -pc_factor_mat_solver_type'
+  petsc_options_value = 'lu      300                superlu_dist'
   start_time = 0.0
   end_time = 10.0
   dt = 1.0
```

**Reply by dschwen** [reply]
When you use the `material_properties` argument you need to be careful, when those properties depend on non-linear variables. To ensurer propper chain rule derivatives read up here: https://mooseframework.inl.gov/moose/source/materials/DerivativeParsedMaterial.html the section following

> The ["material_property_names"](https://mooseframework.inl.gov/moose/source/materials/DerivativeParsedMaterial.html#6c67f538-74ca-442e-9613-308268765c85) are parsed by the [FunctionMaterialPropertyDescriptor class](http://mooseframework.org/docs/doxygen/modules/classFunctionMaterialPropertyDescriptor.html), which understands the following syntax:

**Reply by dschwen** [reply]
for example the `cpos` block should look like this

```
  [cpos]
    type = DerivativeParsedMaterial
    property_name = cpos
    material_property_names = 'dfepos(phi) N kT'
    expression = 'N * exp(-dfepos/kT) '
    derivative_order = 2
    coupled_variables = 'phi'
    outputs = other
  []
```

**Reply by lindsayad** [reply]
What if you have nested dependencies, for instance `rhobyeps` depends on `cpos` and `cneg` which depend on `dfeneg/pos` which depend on `phi`? I tried
```
@@ -63,7 +63,7 @@
   [./cpos]
     type = DerivativeParsedMaterial
     property_name = cpos
-    material_property_names = 'dfepos N kT'
+    material_property_names = 'dfepos(phi) N kT'
     expression = 'N * exp(-dfepos/kT) '
     derivative_order = 2
     outputs = other
@@ -71,7 +71,7 @@
   [./cneg]
     type = DerivativeParsedMaterial
     property_name = cneg
-    material_property_names = 'dfeneg N kT'
+    material_property_names = 'dfeneg(phi) N kT'
     expression = 'N * exp(-dfeneg/kT) '
     derivative_order = 2
     outputs = other
@@ -79,7 +79,7 @@
   [./rhobyeps]
     type = DerivativeParsedMaterial
     property_name = 'rhobyeps'
-    material_property_names = 'cpos cneg e eps'
+    material_property_names = 'cpos(phi) cneg(phi) e eps'
     expression = 'e*(cpos - cneg)/eps'
     derivative_order = 2
     outputs = other
```
and the Jacobian is still not good

**Reply by lindsayad** [reply]
Oops, I hadn't added `coupled_variables`. With that addition, the Jacobian now looks perfect

**Reply by lindsayad** [reply, marked as answer]
So in summary, the working patch is
```
diff --git a/modules/phase_field/user-problem.i b/modules/phase_field/user-problem.i
index 16bb511818..6878f02beb 100644
--- a/modules/phase_field/user-problem.i
+++ b/modules/phase_field/user-problem.i
@@ -63,25 +63,28 @@
   [./cpos]
     type = DerivativeParsedMaterial
     property_name = cpos
-    material_property_names = 'dfepos N kT'
+    material_property_names = 'dfepos(phi) N kT'
     expression = 'N * exp(-dfepos/kT) '
     derivative_order = 2
+    coupled_variables = 'phi'
     outputs = other
   [../]
   [./cneg]
     type = DerivativeParsedMaterial
     property_name = cneg
-    material_property_names = 'dfeneg N kT'
+    material_property_names = 'dfeneg(phi) N kT'
     expression = 'N * exp(-dfeneg/kT) '
     derivative_order = 2
     outputs = other
+    coupled_variables = 'phi'
   [../]
   [./rhobyeps]
     type = DerivativeParsedMaterial
     property_name = 'rhobyeps'
-    material_property_names = 'cpos cneg e eps'
+    material_property_names = 'cpos(phi) cneg(phi) e eps'
     expression = 'e*(cpos - cneg)/eps'
     derivative_order = 2
     outputs = other
+    coupled_variables = 'phi'
   [../]
 []

@@ -119,18 +122,18 @@
   type = Transient
   scheme = bdf2
   verbose = True
   solve_type = 'Newton'
-  l_max_its = 50
-  l_tol = 1e-6
-  nl_max_its = 50
+  l_max_its = 300
+  l_tol = 1e-2
+  nl_max_its = 30
   nl_rel_tol = 1e-8
   nl_abs_tol = 1e-12
 #  petsc_options = '-snes_monitor -ksp_monitor_true_residual -snes_converged_reason -ksp_converged_reason'
 #  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
 #  petsc_options_value = 'asm      31                  preonly      lu          4'
 #  line_search = 'none'
-  petsc_options_iname = -pc_type
-  petsc_options_value = asm
+  petsc_options_iname = '-pc_type -ksp_gmres_restart -pc_factor_mat_solver_type'
+  petsc_options_value = 'lu	  300		     superlu_dist'
   start_time = 0.0
   end_time = 10.0
   dt = 1.0
```
